### ISSUE TITLE
Error: error: no matching function for call to ‘sort<...>(...)’

### ISSUE THREAD

**Issue by pbertoni89** [shwina/stdpar-cython#1]
Just forked and testing in Ubuntu 20.04 with Nvidia Quadro K2200, Cuda 11.0, HPC-SDK v20.9.

Error persists in all environments: g++9, g++10; python 3.6, python 3.8.

Within `sort.ipynb` context, just after writing `cppsort_parallel.pyx`, setup
```
# %%capture
!python setup.py build_ext --inplace
```
Fails with error code 1
```
Compiling cppsort_parallel.pyx because it changed.
[1/1] Cythonizing cppsort_parallel.pyx
/media/xnext/DATAEXT/opt/miniconda/envs/ds38/lib/python3.8/site-packages/Cython/Compiler/Main.py:369: FutureWarning: Cython directive 'language_level' not set, using 2 for now (Py2). This will change in a later release! File: /media/xnext/DATAEXT/git_repos/gpustdparpy/cppsort_parallel.pyx
  tree = Parsing.p_module(s, pxd, full_module_name)
running build_ext
building 'cppsort_parallel' extension
gcc -pthread -B /media/xnext/DATAEXT/opt/miniconda/envs/ds38/compiler_compat -Wl,--sysroot=/ -Wsign-compare -DNDEBUG -g -fwrapv -O3 -Wall -Wstrict-prototypes -fPIC -I/media/xnext/DATAEXT/opt/miniconda/envs/ds38/include/python3.8 -c cppsort_parallel.cpp -o build/temp.linux-x86_64-3.8/cppsort_parallel.o -std=c++17
cc1plus: warning: command-line option ‘-Wstrict-prototypes’ is valid for C/ObjC but not for C++
cppsort_parallel.cpp: In function ‘PyObject* __pyx_pf_16cppsort_parallel_cppsort(PyObject*, __Pyx_memviewslice)’:
cppsort_parallel.cpp:2159:256: error: no matching function for call to ‘sort<const __pstl::execution::v1::parallel_policy, int*>(const __pstl::execution::v1::parallel_policy&, int*, int*)’
 2159 | im=0 */ (__pyx_v_x.data + __pyx_t_3 * __pyx_v_x.strides[0]) )))) + 1));
      |                                                                      ^
In file included from /usr/include/c++/10/functional:65,
                 from /usr/include/c++/10/pstl/algorithm_impl.h:16,
                 from /usr/include/c++/10/pstl/glue_execution_defs.h:50,
                 from /usr/include/c++/10/execution:32,
                 from cppsort_parallel.cpp:642:
/usr/include/c++/10/bits/stl_algo.h:4849:5: note: candidate: ‘template<class _RAIter> void std::sort(_RAIter, _RAIter)’
 4849 |     sort(_RandomAccessIterator __first, _RandomAccessIterator __last)
      |     ^~~~
/usr/include/c++/10/bits/stl_algo.h:4849:5: note:   template argument deduction/substitution failed:
cppsort_parallel.cpp:2159:256: error: wrong number of template arguments (2, should be 1)
 2159 | im=0 */ (__pyx_v_x.data + __pyx_t_3 * __pyx_v_x.strides[0]) )))) + 1));
      |                                                                      ^
In file included from /usr/include/c++/10/functional:65,
                 from /usr/include/c++/10/pstl/algorithm_impl.h:16,
                 from /usr/include/c++/10/pstl/glue_execution_defs.h:50,
                 from /usr/include/c++/10/execution:32,
                 from cppsort_parallel.cpp:642:
/usr/include/c++/10/bits/stl_algo.h:4880:5: note: candidate: ‘void std::sort(_RAIter, _RAIter, _Compare) [with _RAIter = const __pstl::execution::v1::parallel_policy; _Compare = int*]’
 4880 |     sort(_RandomAccessIterator __first, _RandomAccessIterator __last,
      |     ^~~~
/usr/include/c++/10/bits/stl_algo.h:4880:63: note:   no known conversion for argument 2 from ‘int*’ to ‘__pstl::execution::v1::parallel_policy’
 4880 |     sort(_RandomAccessIterator __first, _RandomAccessIterator __last,
      |                                         ~~~~~~~~~~~~~~~~~~~~~~^~~~~~
In file included from /usr/include/c++/10/execution:38,
                 from cppsort_parallel.cpp:642:
/usr/include/c++/10/pstl/glue_algorithm_impl.h:640:1: note: candidate: ‘template<class _ExecutionPolicy, class _RandomAccessIterator, class _Compare> __pstl::__internal::__enable_if_execution_policy<_ExecutionPolicy, void> std::sort(_ExecutionPolicy&&, _RandomAccessIterator, _RandomAccessIterator, _Compare)’
  640 | sort(_ExecutionPolicy&& __exec, _RandomAccessIterator __first, _RandomAccessIterator __last, _Compare __comp)
      | ^~~~
/usr/include/c++/10/pstl/glue_algorithm_impl.h:640:1: note:   template argument deduction/substitution failed:
cppsort_parallel.cpp:2159:256: note:   candidate expects 4 arguments, 3 provided
 2159 | im=0 */ (__pyx_v_x.data + __pyx_t_3 * __pyx_v_x.strides[0]) )))) + 1));
      |                                                                      ^
 In file included from /usr/include/c++/10/execution:38,
                 from cppsort_parallel.cpp:642:
/usr/include/c++/10/pstl/glue_algorithm_impl.h:653:1: note: candidate: ‘__pstl::__internal::__enable_if_execution_policy<_ExecutionPolicy, void> std::sort(_ExecutionPolicy&&, _RandomAccessIterator, _RandomAccessIterator) [with _ExecutionPolicy = const __pstl::execution::v1::parallel_policy; _RandomAccessIterator = int*; __pstl::__internal::__enable_if_execution_policy<_ExecutionPolicy, void> = void]’ (near match)
  653 | sort(_ExecutionPolicy&& __exec, _RandomAccessIterator __first, _RandomAccessIterator __last)
      | ^~~~
/usr/include/c++/10/pstl/glue_algorithm_impl.h:653:1: note:   conversion of argument 1 would be ill-formed:
cppsort_parallel.cpp:2159:77: error: cannot bind rvalue reference of type ‘const __pstl::execution::v1::parallel_policy&&’ to lvalue of type ‘const __pstl::execution::v1::parallel_policy’
 2159 | d::execution::parallel_policy const ,int *>(std::execution::par, (&(*((int *) ( /* dim=0 */ (__pyx_v_x.data + __pyx_t_1 * __pyx_v_x.strides[0]) )))), ((&(*((int *) ( /* dim=0 */ (__pyx_v_x.data + __pyx_t_3 * __pyx_v_x.strides[0]) )))) + 1));
      |                                             ~~~~~~~~~~~~~~~~^~~
error: command 'gcc' failed with exit status 1
```

**Comment by shwina** [comment]
Thanks for reporting! Are you able to provide the Cython generated C++ file `cppsort_parallel.cpp`?

**Comment by pbertoni89** [comment]
Quite huge, of course. Zipping and attaching for convenience.
[cppsort_parallel.zip](https://github.com/shwina/stdpar-cython/files/5567001/cppsort_parallel.zip)

**TL; DR**: it seems the `Compare comp` argument is missing ([cppreference.com](https://en.cppreference.com/w/cpp/algorithm/sort))

**Comment by pbertoni89** [comment]
Guess it's related to a line inside `algorithm.pxd`: there are 3 arguments instead of 4 expected by the standard

```
void sort[ExecutionPolicy, Iter](ExecutionPolicy&& policy, Iter first, Iter last) except +
```

**Comment by shwina** [comment]
Hmm - the specific overload that refers to is (2) [here](https://en.cppreference.com/w/cpp/algorithm/sort):

```
template< class ExecutionPolicy, class RandomIt >
void sort( ExecutionPolicy&& policy, RandomIt first, RandomIt last );
```

**Comment by shwina** [comment]
So the problem is in line 2159 in the template params:

```c++
std::sort<std::execution::parallel_policy const ,int *>
```

where the template params should actually be (and I get the following):

```c++
std::sort<std::execution::parallel_policy const &,int *>
```

May I ask what version of Cython you are using to generate the C++?

**Comment by shwina** [comment]
I can reproduce the issue you're seeing with Cython 0.29. To run these notebooks you will need a development version of Cython as that has support for universal references. One way to get that is to use`pip`:

```
python -m pip install git+https://github.com/cython/cython
```

**Comment by pbertoni89** [comment]
May I be damned. Now it works, thanks!!!
